lib/Conversion/CalyxToHW/CalyxToHW.cpp

```diff
@@ -42,40 +42,36 @@ struct ConvertComponentOp : public OpConversionPattern<ComponentOp> {
   matchAndRewrite(ComponentOp component, OpAdaptor adaptor,
                   ConversionPatternRewriter &rewriter) const override {
     SmallVector<hw::PortInfo> hwInputInfo;
-    for (auto [name, type, direction, _] : component.getPortInfo())
+    auto portInfo = component.getPortInfo();
+    for (auto [name, type, direction, _] : portInfo)
       hwInputInfo.push_back({name, hwDirection(direction), type});
-
-    auto hwMod = rewriter.create<HWModuleOp>(
-        component.getLoc(), component.getNameAttr(), hwInputInfo);
-
-    rewriter.eraseOp(hwMod.getBodyBlock()->getTerminator());
-    ConversionPatternRewriter::InsertionGuard g(rewriter);
-    rewriter.setInsertionPointToEnd(hwMod.getBodyBlock());
+    ModulePortInfo hwPortInfo(hwInputInfo);
 
     SmallVector<Value> argValues;
-    SmallVector<Value> outputWires;
-    size_t portIdx = 0;
-    for (auto [name, type, direction, _] : component.getPortInfo()) {
-      switch (direction) {
-      case calyx::Direction::Input:
-        assert(hwMod.getArgument(portIdx).getType() == type);
-        argValues.push_back(hwMod.getArgument(portIdx));
-        break;
-      case calyx::Direction::Output:
-        auto wire = rewriter.create<sv::WireOp>(component.getLoc(), type, name);
-        auto wireRead =
-            rewriter.create<sv::ReadInOutOp>(component.getLoc(), wire);
-        argValues.push_back(wireRead);
-        outputWires.push_back(wireRead);
-        break;
-      }
-      ++portIdx;
-    }
-
+    auto hwMod = rewriter.create<HWModuleOp>(
+        component.getLoc(), component.getNameAttr(), hwPortInfo,
+        [&](OpBuilder &b, HWModulePortAccessor &ports) {
+          for (auto [name, type, direction, _] : portInfo) {
+            switch (direction) {
+            case calyx::Direction::Input:
+              assert(ports.getInput(name).getType() == type);
+              argValues.push_back(ports.getInput(name));
+              break;
+            case calyx::Direction::Output:
+              auto wire = b.create<sv::WireOp>(component.getLoc(), type, name);
+              auto wireRead =
+                  b.create<sv::ReadInOutOp>(component.getLoc(), wire);
+              argValues.push_back(wireRead);
+              ports.setOutput(name, wireRead);
+              break;
+            }
+          }
+        });
+
+    auto *outputOp = hwMod.getBodyBlock()->getTerminator();
     rewriter.mergeBlocks(component.getBody(), hwMod.getBodyBlock(), argValues);
-    rewriter.create<OutputOp>(component.getLoc(), outputWires);
+    outputOp->moveAfter(&hwMod.getBodyBlock()->back());
```

Comment:
Why is this necessary?

Reply:
I assume it's because the builder used to create the HWModuleOp now creates it automatically? I think this is generally unsafe, and the mutation here should be wrapped in `rewriter.updateRootInPlace(hwMod, ...)`

Reply:
Also, for the record, I don't think it's safe to use `b` in the body of the lambda that creates the HWModuleOp. I'm pretty sure all IR mutation needs to happen through `rewriter`